std-bits/table/src/main/java/org/enso/table/data/column/storage/ObjectStorage.java

```diff
@@ -40,15 +42,29 @@ public StorageType getType() {
   public StorageType inferPreciseType() {
     if (inferredType == null) {
       StorageType currentType = null;
+
       for (int i = 0; i < size(); i++) {
         var item = getItemBoxed(i);
-        if (item != null) {
-          var itemType = StorageType.forBoxedItem(item);
-          currentType = StorageType.findCommonType(currentType, itemType);
-          if (currentType instanceof AnyObjectType) {
-            break;
+        if (item == null) {
+          continue;
+        }
+
+        var itemType = StorageType.forBoxedItem(item);
+        if (currentType == null) {
+          currentType = itemType;
+        } else if (currentType != itemType) {
+          // Allow mixed integer and float types in a column, returning a float.
+          if ((itemType instanceof IntegerType && currentType instanceof FloatType)
+              || (itemType instanceof FloatType && currentType instanceof IntegerType)) {
+            currentType = FloatType.FLOAT_64;
+          } else {
+            currentType = AnyObjectType.INSTANCE;
           }
         }
+
+        if (currentType instanceof AnyObjectType) {
+          break;
+        }
```

Comment:
I like the early quit - for many storages it will mean that computing this will finish very quickly.